src/reanimated2/Hooks.js

```diff
@@ -304,36 +304,41 @@ export function useAnimatedGestureHandler(handlers) {
   return useEvent(
     event => {
       'worklet';
-      const UNDETERMINED = 0;
       const FAILED = 1;
+      const BEGAN = 2;
       const CANCELLED = 3;
       const ACTIVE = 4;
       const END = 5;
 
-      if (event.oldState === UNDETERMINED && handlers.onStart) {
+      if (event.state === BEGAN && handlers.onStart) {
         handlers.onStart(event, context);
       }
       if (event.state === ACTIVE && handlers.onActive) {
         handlers.onActive(event, context);
       }
-      if (event.oldState === ACTIVE && event.state === END && handlers.onEnd) {
+      if (
+        (event.oldState === BEGAN || event.oldState === ACTIVE) &&
+        event.state === END && handlers.onEnd
+      ) {
         handlers.onEnd(event, context);
       }
       if (
-        event.oldState === ACTIVE &&
-        event.state === FAILED &&
-        handlers.onFail
+        (event.oldState === BEGAN || event.oldState === ACTIVE) &&
+        event.state === FAILED && handlers.onFail
       ) {
         handlers.onFail(event, context);
       }
       if (
-        event.oldState === ACTIVE &&
-        event.state === CANCELLED &&
-        handlers.onCancel
+        (event.oldState === BEGAN || event.oldState === ACTIVE) &&
```

Comment:
And according again to this flow it should check for active only https://docs.swmansion.com/react-native-gesture-handler/docs/state.html#state-flows

Reply:
The way I interpreted the docs was that those were some of the few possible flows, but there could be more possible flows depending on the situation. But I think you are probably right.
 
Also I mentioned this in the PR that I was just being conservative in checking the conditions, so I could get a proper review.

So i'll go ahead with working on all of your suggestions.